hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/OzoneManagerLock.java

```diff
@@ -154,79 +178,137 @@ public void releaseVolumeLock(String volume) {
+    myLocks.get().get(BUCKET_LOCK).decrementAndGet();
   }
 
   /**
-   * Acquires bucket lock on the given resource.
+   * Acquires user lock on the given resource.
    *
    * <p>If the lock is not available then the current thread becomes
    * disabled for thread scheduling purposes and lies dormant until the
    * lock has been acquired.
    *
-   * @param bucket Bucket on which the lock has to be acquired
+   * @param user User on which the lock has to be acquired
    */
-  public void acquireBucketLock(String volume, String bucket) {
-    manager.lock(OM_KEY_PREFIX + volume + OM_KEY_PREFIX + bucket);
-    myLocks.get().get(BUCKET_LOCK).incrementAndGet();
+  public void acquireUserLock(String user) {
+    // In order to not maintain username's on which we have acquired lock,
+    // just checking have we acquired userLock before. If user want's to
+    // acquire user lock on multiple user's they should use
+    // acquireMultiUserLock. This is just a protection logic, to let not users
+    // use this if acquiring lock on multiple users. As currently, we have only
+    // use case we have for this is during setOwner operation in VolumeManager.
+    if (hasAnyUserLock()) {
+      LOG.error("Already have userLock");
+      throw new RuntimeException("For acquiring lock on multiple users, use " +
+          "acquireMultiLock method");
+    }
+    manager.lock(OM_USER_PREFIX + user);
+    myLocks.get().get(USER_LOCK).incrementAndGet();
   }
 
   /**
-   * Releases the bucket lock on given resource.
+   * Releases the user lock on given resource.
    */
-  public void releaseBucketLock(String volume, String bucket) {
-    manager.unlock(OM_KEY_PREFIX + volume + OM_KEY_PREFIX + bucket);
-    myLocks.get().get(BUCKET_LOCK).decrementAndGet();
+  public void releaseUserLock(String user) {
+    manager.unlock(OM_USER_PREFIX + user);
+    myLocks.get().get(USER_LOCK).decrementAndGet();
   }
 
   /**
-   * Returns true if the current thread holds any volume lock.
-   * @return true if current thread holds volume lock, else false
+   * Acquire user lock on 2 users. In this case, we compare 2 strings
+   * lexicographically, and acquire the locks according to the sorted order of
+   * the user names. In this way, when acquiring locks on multiple user's, we
+   * can avoid dead locks. This method should be called when single thread is
+   * acquiring lock on 2 users at a time.
+   *
+   * Example:
+   * ozone, hdfs -> lock acquire order will be hdfs, ozone
+   * hdfs, ozone -> lock acquire order will be hdfs, ozone
+   *
+   * @param newUser
+   * @param oldUser
    */
-  private boolean hasAnyVolumeLock() {
-    return myLocks.get().get(VOLUME_LOCK).get() != 0;
+  public void acquireMultiUserLock(String newUser, String oldUser) {
+    Preconditions.checkNotNull(newUser);
+    Preconditions.checkNotNull(oldUser);
+    int compare = newUser.compareTo(oldUser);
+
+    if (compare < 0) {
+      manager.lock(OM_USER_PREFIX + newUser);
```

Comment:
Sorry not understood why it will raise to dead lock. As already thread acquired locks on anu, bharat. The second time we increment thread-local variables, and increment count of the active lock. Sorry if I am missing something here.

Reply:
Lock Object is re-entrant, there fore should not be an issue.